_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+@EnableFeignClients
+@EnableDiscoveryClient
+@SpringBootApplication
+public class ApiGatewayApplication {
+
+    public static void main(String[] args) {
+        SpringApplication.run(ApiGatewayApplication.class, args);
+    }
+}
+
+record Car(String name) {
+}
+
+@FeignClient("car-service")
+interface CarClient {
+
+    @GetMapping("/cars")
+    CollectionModel<Car> readCars();
+}
+
+@RestController
+class CoolCarController {
+
+    private final CarClient carClient;
+
+    public CoolCarController(CarClient carClient) {
+        this.carClient = carClient;
+    }
+
+    @GetMapping("/cool-cars")
+    public Collection<Car> coolCars() {
+        return carClient.readCars()
+                .getContent()
+                .stream()
+                .filter(this::isCool)
+                .collect(Collectors.toList());
+    }
+
+    private boolean isCool(Car car) {
+        return !car.name().equals("AMC Gremlin") &&
+                !car.name().equals("Triumph Stag") &&
+                !car.name().equals("Ford Pinto") &&
+                !car.name().equals("Yugo GV");
+    }
+}
+```
+
+This worked great, but I still wanted to proxy `/home` to the downstream car service.
+
+## Add Routing with Spring Cloud Gateway
+
+I immediately discovered that adding `spring-cloud-starter-gateway` as a dependency caused issues. First, I had Spring MVC in my classpath, and Spring Cloud Gateway uses WebFlux. WebFlux recommends using WebClient over Feign and Resilience4J over Hystrix. I decided to switch to WebClient and Resilience4J.
+
+I had to remove the following dependencies from my original `api-gateway` project.
+
+```groovy
+implementation 'org.springframework.boot:spring-boot-starter-data-rest'
+implementation 'org.springframework.boot:spring-boot-starter-web'
+implementation 'org.springframework.cloud:spring-cloud-starter-openfeign'
+```
+
+And add Spring Cloud Gateway with Resilience4j dependencies:
+
+```groovy
+implementation 'org.springframework.cloud:spring-cloud-starter-circuitbreaker-reactor-resilience4j'
+implementation 'org.springframework.cloud:spring-cloud-starter-gateway'
+```
+
+Then, I moved `CoolCarController` to its own class and re-implemented it with WebClient.
+
+```java
+package com.example.apigateway.web;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.client.circuitbreaker.ReactiveCircuitBreaker;
+import org.springframework.cloud.client.circuitbreaker.ReactiveCircuitBreakerFactory;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Flux;
+
+@RestController
+class CoolCarController {
+
+    Logger log = LoggerFactory.getLogger(CoolCarController.class);
+
+    private final WebClient.Builder webClientBuilder;
```

Comment:
I think worth mentioneing that a bean needs to be configured for this to work

Reply:
I'll add this in a commit since I can't do code inside a suggestion.

Reply:
Added below.